In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import seaborn as sns
import plotly.offline as py
from plotly.offline import init_notebook_mode, iplot
import plotly.graph_objs as go
from plotly import tools
init_notebook_mode(connected=True)  
import plotly.figure_factory as ff
# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

['blockchai', 'blockc']


In [2]:
df = pd.read_csv('../input/blockchai/crypto.csv')
df.head()

Market cap  close       date  high   low  open      volume Currency
0  3526468020   3.89  22-Feb-19  3.90  3.77  3.80  1413643823      EOS
1  3449084067   3.81  21-Feb-19  3.93  3.73  3.93  1392322652      EOS
2  3553278085   3.92  20-Feb-19  3.92  3.52  3.55  1914563883      EOS
3  3231790105   3.57  19-Feb-19  3.79  3.49  3.51  2028171138      EOS
4  3144033218   3.47  18-Feb-19  3.50  2.87  2.88  1867708700      EOS

In [3]:
df.tail()

Market cap  close       date    ...      open    volume  Currency
1739           0   5.63  31-Jan-19    ...      6.97   4994843      GRIN
1740           0   6.93  30-Jan-19    ...      8.88   5777397      GRIN
1741           0   8.84  29-Jan-19    ...     10.34   6310407      GRIN
1742           0  10.31  28-Jan-19    ...     14.87  10613946      GRIN
1743           0  14.36  27-Jan-19    ...      9.96  15941428      GRIN

[5 rows x 8 columns]

In [4]:
df['date'] = pd.to_datetime(df['date'])

In [5]:
df.dtypes

Market cap             int64
close                float64
date          datetime64[ns]
high                 float64
low                  float64
open                 float64
volume                 int64
Currency              object
dtype: object

Market vs. Currency
Part of deciding which cryptocurrencies are to invest in or predicting the prices of cryptocurrencies is determining how much value to place on the technical analysis of the cryptocurrency itself an how much value to place on analysis of the market.

When cryptocurrencies are moving as a result of a bullish or bearish market, they tend to be moving in the same direction at a similar pace. This means that if most of the cryptocurrencies are moving similarly, it is more likely that this movement is related to the market more than it is the investment potential (or FOMO, FUD or whatever other reason people have for buying or selling) of the individual cryptocurrency. So in this notebook, we look at quantifying the similarities/differences in the movements of cryptocurrencies.

Once we have an idea of how various cryptocurrencies are tending to move, we can make a more informed decision on what to invest in, based on how much risk we are comfortable with and how much returns we can expect. As modern portfolio theory states, higher risk is associated with greater probability of higher return and lower risk with a greater probability of smaller return.

In the first sections we will examine daily price changes and variance. In further sections we will build a basic market index for cryptocurrencies, and also a volatility index (not yet implemented) and see how it compares to regular stock market volatility.

Lastly (not yet implemented), I would like to examine how to put together a cryptocurrency portfolio based on optimising for return and minimising risk (as far as possible with cryptos!).

# Types of cryptocurrency

The blockchain brings together the three main types of cryptocurrency. Bitcoin was the first blockchain (we aren't going to discuss about bitcoin on how it started and what it does).
After Bitcoin, many new blockchains were created — these are called altcoins. NEO, Litecoin and Cardano are solid examples of altcoins.  Finally, I must introduce you to tokens/dApps — the third main type of cryptocurrencies. Examples of these include EOS, AUGUR, and WePower (WPR).

There are various types of cryptocurrency but i'm going to be specific about some special types of cryptocurrency which aren't bitcoin but very demanding.

1- Programming currency- EOS & AUGUR

2- Privacy based currency - GRIN

# EOS analysis

EOSIO adopts a free-market approach to allocating scarce resources to their highest purpose. To facilitate this market, the eosio system contract allows users to buy RAM from the system and sell RAM back to the system in exchange for the blockchains native tokens (e.g EOS). This provides liquidity in the RAM market while facilitating price discovery. The less unallocated RAM available to the market maker the higher the market maker prices the remaining RAM. The algorithm used for this market maker is known as a Bancor Relay. A Bancor Relay does not set the price of RAM but rather offers to buy and sell at previously established market rates. Anytime the current market rate is different than the current price offered by the Bancor Relay, traders will buy or sell RAM accordingly, reducing the market determined price gap.

In [6]:
fig = ff.create_candlestick(df.open[df.Currency == 'EOS'], df.high[df.Currency == 'EOS'], df.low[df.Currency == 'EOS'], df.close[df.Currency == 'EOS'], dates=df.date[df.Currency == 'EOS'])
fig['layout'].update({
    'title': 'EOS price development',
    'yaxis': {'title': 'EOS in USD'}
})
py.iplot(fig)

* Figure shows the historical evolution of the EOS price since February 2018 at around 8.5 USD until May 2018 currently trading just short around 18 USD. In the span of less than three months, it has roughly doubled in value in anticipation of the main net, reaching an all-time high of 23.3 USD on April 29th.
* As it is the case with the vast majority of cryptocurrencies, there is a clear correlation between the price trends of EOS and those of Bitcoin and particularly Ethereum, with which it shares a similar working concept. April was an extremely profitable month for EOS investors, with a surge in price of over 300 percent in just a few weeks. The price has since experienced a pullback and is now quite stable within the 3 - 4 USD range.

The 2018 cryptocurrency crash(also known as the Bitcoin crash and the Great crypto crash) was the sell-off of most cryptocurrencies from January 2018. After an unprecedented boom in 2017, the price of bitcoin fell by about 65 percent during the month from 6 January to 6 February 2018. Subsequently, nearly all other cryptocurrencies also peaked from December 2017 through January 2018, and then followed bitcoin. The cryptocurrencies' market capitalization lost at least 342 billion US dollars in the first quarter of 2018, the largest loss in cryptocurrencies up to that date. By September 2018, cryptocurrencies collapsed 80% from their peak in January 2018, making the 2018 cryptocurrency crash worse than the Dot-com bubble's 78% collapse. By 26 November, bitcoin also fell by over 80% from its peak, having lost almost one-third of its value in the previous week. 


# EOS market analysis

In [7]:
df2 = pd.read_csv('../input/blockc/eos_market.csv')
df2.head()

pair  price    source     volume  volume%
0   EOS/ETH   3.50     Bibox  132293861     8.52
1  EOS/USDT   3.51      OKEx  127507656     8.21
2  EOS/USDT   3.51    ZB.COM   83477607     5.38
3   EOS/BTC   3.51      OKEx   58949705     3.80
4   EOS/BTC   3.51  RightBTC   58394758     3.76

In [8]:
import random
def random_colors(number_of_colors):
    color = ["#"+''.join([random.choice('0123456789ABCDEF') for j in range(6)])
                 for i in range(number_of_colors)]
    return color

In [9]:
trace0 = go.Scatter(
    x=df2['volume%'],
    y=df2.volume,
    text=df2.source,
    mode='markers',
    marker=dict(
        color= random_colors(20),
        size=[100, 80, 60, 40],
    )
)
layout = {
    'title': 'EOS market source ',
    'yaxis': {'title': 'EOS volume in USD'},
    'xaxis': {'title': 'EOS volume percentage'}
}

data = [trace0]
fig = dict(data=data, layout=layout)
py.iplot(fig, filename='bubblechart-text')

EOS main market source is Bitbox whose transaction volume is 132 million USD, it also occupies 8.52 volume percentage of the total transaction volume, it's the biggest transaction share Bitbox has with EOS.
Follow up is OKEX which has almost 127 million USD transaction volume in EOS currency while occupying 8.21 percentage of the total transaction volume, it tells us about the market share of OKEX has been dealt in EOS.

In [10]:
trace0 = go.Scatter(
    x=df2.price,
    y=df2.volume,
    text=df2.pair,
    mode='markers',
    marker=dict(
        color= random_colors(20),
        size=[100, 80, 60, 40],
    )
)
layout = {
    'title': 'EOS market transaction pairs ',
    'yaxis': {'title': 'EOS volume in USD'},
    'xaxis': {'title': 'EOS price in USD'}
}

data = [trace0]
fig = dict(data=data, layout=layout)
py.iplot(fig, filename='bubblechart-text')

Bitbox main transaction pair is EOS/ETH whose transaction volume contains 132 million USD, it also occupies 3.52 average price of the total transaction volume, it's the biggest transaction share Bitbox has with EOS/ETH.
Follow up is OKEX which has almost 127 million USD transaction volume in EOS currency while having 3.51 average price of the total transaction volume, it tells us about the market share of OKEX has been dealt in EOS/USD.

In [11]:
# Get daily percent change in close price
df['pct_change'] = df['close'].groupby(level=0).pct_change()
df.head()

Market cap  close       date     ...          volume  Currency  pct_change
0  3526468020   3.89 2019-02-22     ...      1413643823       EOS         NaN
1  3449084067   3.81 2019-02-21     ...      1392322652       EOS   -0.020566
2  3553278085   3.92 2019-02-20     ...      1914563883       EOS    0.028871
3  3231790105   3.57 2019-02-19     ...      2028171138       EOS   -0.089286
4  3144033218   3.47 2019-02-18     ...      1867708700       EOS   -0.028011

[5 rows x 9 columns]

In [12]:
data = [go.Scatter(
          x=df['date'].where(df['Currency'] == 'EOS'),
          y=df['pct_change'].where(df['Currency'] == 'EOS')
)]
layout = dict(
    title = "Percent change in closing value of EOS"
)
fig = dict(data=data, layout=layout)
py.iplot(fig)

 Percentage change in closing price of EOS has been variant while most of them lying arounf 0.1 but occasionaly it spiked upto 0.46 especially on Sep 3 2017, it tells us about the market boom that EOS had due to Bitcoin spark into the market.
 Consistent negative drop can be seen after May 2018 due to crypto market crash which brought tragedy on all currency closing values, but ever since EOS has recovered well enought to get that average 0.1 percent change recently.

# AUGUR(REP) analysis

Augur is one of the few Dapps live on the Ethereum main net, where it launched in July of this year. It has been one of the most popular projects in the Ethereum space from the early beginning, being able to maintain most of its momentum since then. Other prediction markets are being built in the ecosystem but none has had the levels of early product usage and community activity that Augur has enjoyed so far.

Almost half a year after its launch, in this post we dive into Augur’s prediction market functionality, provide a number of relevant metrics about its usage throughout these months, and explore its limitations in areas that need improvement.

In [13]:
trace = go.Ohlc(x=df.date[df.Currency == 'AUGUR'],
                open=df.open[df.Currency == 'AUGUR'],
                high=df.high[df.Currency == 'AUGUR'],
                low=df.low[df.Currency == 'AUGUR'],
                close=df.close[df.Currency == 'AUGUR']
               )  
data = [trace]         
layout = {
    'title': 'AUGUR price development ',
    'yaxis': {'title': 'AUGUR price in USD'},
}
fig = dict(data=data, layout=layout)
py.iplot(fig, filename='simple_candlestick')

Market analysis

Augur (REP) is a cryptocurrency or a form of digital asset. Augur (REP) price for today is 12.61 USD with a 24-hour trading volume of  3,472,934 USD. Price is down -1.3% in the last 24 hours. It has a circulating supply of 11 Million coins and a max supply of 11 Million coins.

Technical analysis: 

. AUGUR/BINANCECOIN is in a range bound and the beginning of uptrend is expected. 

. The price is below the 21-Day WEMA which acts as a dynamic resistance. 

. The RSI is at 40. 

. The price downtrend in the daily chart and the RSI downtrend are broken, so the probability of the resumption of an uptrend is increased. 

Trading suggestion:

. There is a possibility of temporary retracement to suggested price zone (13.13 to 14.20). if so, traders can set orders based on Price Action and expect to reach short-term targets. 



In [14]:
data = [go.Scatter(
          x=df['date'].where(df['Currency'] == 'AUGUR'),
          y=df['pct_change'].where(df['Currency'] == 'AUGUR')
)]
layout = dict(
    title = "Percent change in closing value of AUGUR"
)
fig = dict(data=data, layout=layout)
py.iplot(fig)

 Percentage change in closing price of AUGUR has been variant while most of them lying arounf 0.1 but occasionaly it spiked upto 0.2 or 0.3 especially on Sep 3 2016, it tells us about the market boom that AUGUR had due to reintroduction into the market.
 Consistent negative drop can be seen after May 2018 due to crypto market crash which brought tragedy on all currency closing values, but ever since AUGUR has recovered well enought to get that average 0.1 percent change recently.

# GRIN analysis

Debuted on January 15, 2019, GRIN is a cryptocurrency implementation of MimbleWimble created to provide privacy, scalability and fungibility. Explaining GRIN from a technical perspective, it contains substantial overlap with the previously mentioned MimbleWimble.

Thus, it’s good to emphasize other components like a consensus as well as monetary policy.

GRIN retains the fungibility and privacy features of MimbleWimble where there aren’t any addresses, transactions amounts, where transactions can be merged removing all intermediary data.

Moreover, blocks in GRIN contain no transactions and the block just seems to be one big transaction.

All spent outputs in GRIN can be safely removed, that allows a drastically reduced size of the blockchain. Users can download and verify the blockchain majorly faster than other digital currencies.

Ultimately, GRIN can scale with the number of users, instead of the number of transactions.

**How does it work?**

While most cryptos work on a principle of full transparency, Grin being a privacy coin looks to do the exact opposite. It has made an effort to ensure that transactions on their chain are anonymous, non-transparent and most importantly untraceable. The Peer to Peer setup allows wallets to trade. It does so by ensuring that only those wallets executing a trade exchange data for a limited time. This can even occur in an asynchronous state and once the transaction is completed all the information is deleted.

Apart from its celebrated opaqueness, the token is also rather scalable. This makes it a much-valued asset. It achieves this by using a novel data storage method, which reorganizes tracking of data blocks and reportedly utilizes more than 75 percent less space than the Bitcoin network. The underlying protocol then compresses the blockchain further by using only a single input and output, eliminating parts of some transactions This not only leads to efficiencies in movement but also hastens the full verification process.

While there are always those who say that if there is nothing to hide then why worry; yet there will always be those who defend privacy as one would defend liberty. As this might be an argument better suited for a socio-political commentary, one can only hope that the opportunities that this coin can offer are understood in the right spirit and implemented accordingly.

In [15]:
trace = go.Ohlc(x=df.date[df.Currency == 'GRIN'],
                open=df.open[df.Currency == 'GRIN'],
                high=df.high[df.Currency == 'GRIN'],
                low=df.low[df.Currency == 'GRIN'],
                close=df.close[df.Currency == 'GRIN']
               )  
data = [trace]         
layout = {
    'title': 'GRIN price development ',
    'yaxis': {'title': 'GRIN price in USD'},
}
fig = dict(data=data, layout=layout)
py.iplot(fig, filename='simple_candlestick')

Market analysis

* Figure shows that introduction of the GRIN price since January 2019 at around 14.5 USD until February 2019 currently trading just short around 4 USD. In the span of less than three months, it has roughly one fourth in value in anticipation of the main net, reaching an all-time high of 14.6 USD on Jan 10th.
* As it is the case with the vast majority of cryptocurrencies, there is a clear correlation between the price trends of GRIN and those of Bitcoin and particularly Monero, with which it shares a similar working concept. February was an extremely profitable month for GRIN investors, with a stability in price of currency after deflation. The price has since experienced a pullback and is now quite stable within the 3 - 4 USD range.


In [16]:
trace0 = go.Scatter(
    x = df.date[df.Currency == 'GRIN'],
    y = df.open[df.Currency == 'GRIN'],
    name = 'open',
    fill='tozeroy',
    connectgaps=True,
    line = dict(
        color = ('rgb(205, 12, 24)'),
        width = 4)
)
trace1 = go.Scatter(
    x = df.date[df.Currency == 'GRIN'],
    y = df.close[df.Currency == 'GRIN'],
    name = 'close',
    fill='tonexty',
    connectgaps=True,
    line = dict(
        color = ('rgb(22, 96, 167)'),
        width = 4,)
)
trace2 = go.Scatter(
    x = df.date[df.Currency == 'GRIN'],
    y = df.high[df.Currency == 'GRIN'],
    name = 'high',
    fill='toself',
    connectgaps=True,
    line = dict(
        color = ('rgb(205, 12, 24)'),
        width = 4,
        dash = 'dash') # dash options include 'dash', 'dot', and 'dashdot'
)

data1 = [trace0, trace1, trace2]

# Edit the layout
layout = dict(title = 'Pointers for GRIN ',
              xaxis = dict(title = 'Time'),
              yaxis = dict(title = 'GRIN in USD'),
              )

fig = dict(data=data1, layout=layout)
py.iplot(fig, filename='styled-line')

The price has since introduction found itself now quite stable within the 4 - 5 USD closing and opening range.